zebrad/src/commands/start.rs

```diff
@@ -601,3 +606,46 @@ impl config::Override<ZebradConfig> for StartCmd {
         Ok(config)
     }
 }
+
+/// Check if there are old database folders and delete them from the filesystem.
+///
+/// Iterate over the files and directories in the databases folder and delete if:
+/// - The state directory exists.
+/// - The entry is a directory.
+/// - The directory name has a lenght of at least 2 characters.
+/// - The directory name has a prefix `v`.
+/// - The directory name without the prefix can be parsed as an unsigned number.
+/// - The parsed number is lower than the hardcoded `DATABASE_FORMAT_VERSION`.
+fn check_and_delete_old_databases(cache_dir: PathBuf) -> Result<(), Report> {
+    let cache_dir = cache_dir.join("state");
+    if cache_dir.exists() {
+        for entry in (cache_dir.read_dir()?).flatten() {
+            if entry.file_type()?.is_dir() {
+                let dir_name = entry
+                    .file_name()
+                    .into_string()
+                    .expect("conversion from osstring to string should work.");
+                if dir_name.len() >= 2 && dir_name.starts_with('v') {
+                    let potential_version_number_string = dir_name
+                        .strip_prefix('v')
+                        .expect("should not panic as we know there is a `v` prefix.")
+                        .to_string();
+                    let version_number: u32 =
+                        potential_version_number_string.parse().unwrap_or(u32::MAX);
+                    if version_number < zebra_state::constants::DATABASE_FORMAT_VERSION {
+                        let delete_path = cache_dir.join(dir_name);
+                        info!(
+                            "deleting {}",
+                            delete_path
+                                .to_str()
+                                .expect("path is valid, should not panic")
```

Comment:
This conversion can fail on some filesystems, depending on the home directory name:
```suggestion
                            "deleting outdated state directory {:?}",
                            delete_path,
```